### ISSUE TITLE
Equalization Loss for Long-Tailed Object Recognition

### ISSUE THREAD

**Issue by KeisukeShimokawa** [KeisukeShimokawa/pytorch-papers-challenge#41]
### 論文へのリンク

[[arXiv:2003.05176] Equalization Loss for Long-Tailed Object Recognition](https://arxiv.org/abs/2003.05176)

### 著者・所属機関

Jingru Tan, Changbao Wang, Buyu Li, Quanquan Li, Wanli Ouyang, Changqing Yin, Junjie Yan

### 投稿日時（YYYY-MM-DD）

2020-03-11

**Comment by KeisukeShimokawa** [comment]
[https://github.com/tztztztztz/eql.detectron2](https://github.com/tztztztztz/eql.detectron2)

**Comment by KeisukeShimokawa** [comment]
```python
# pred_class_logitsの形状は(R, K)であり、1つの画像に対してRは検知した物体の数、Kは背景以外のクラスを示す
self.pred_class_logits = pred_class_logits
```

**Comment by KeisukeShimokawa** [comment]
予測をまず行う。

```python
def predict_probs(self):
    """
    Returns:
        list[Tensor]: A list of Tensors of predicted class probabilities for each image.
            Element i has shape (Ri, K + 1), where Ri is the number of predicted objects
            for image i.
    """
    probs = F.sigmoid(self.pred_class_logits)
    # add a dummy probs, representing a fake score for background class
    n = probs.size(0)
    dummy_probs = probs.new_zeros(n, 1)
    probs = torch.cat([probs, dummy_probs], dim=1)

    return probs.split(self.num_preds_per_image, dim=0)
```

E(r)を計算する

```python
def exclude_func(self):
    # instance-level weight
    bg_ind = self.n_c
    weight = (self.gt_classes != bg_ind).float()
    weight = weight.view(self.n_i, 1).expand(self.n_i, self.n_c)
    return weight
```

Tλ(fj)を計算する

```python
def threshold_func(self):
    # class-level weight
    weight = self.pred_class_logits.new_zeros(self.n_c)
    weight[self.freq_info < self.lambda_] = 1
    weight = weight.view(1, self.n_c).expand(self.n_i, self.n_c)
    return weight
```

Equalization Lossを計算する

```python
def eql_loss(self):

    self.n_i, self.n_c = self.pred_class_logits.size()

    def expand_label(pred, gt_classes):
        target = pred.new_zeros(self.n_i, self.n_c + 1)
        target[torch.arange(self.n_i), gt_classes] = 1
        return target[:, :self.n_c]

    target = expand_label(self.pred_class_logits, self.gt_classes)

    eql_w = 1 - self.exclude_func() * self.threshold_func() * (1 - target)

    cls_loss = F.binary_cross_entropy_with_logits(self.pred_class_logits, target,
                                                      reduction='none')

    return torch.sum(cls_loss * eql_w) / self.n_i
```
